### ISSUE TITLE
Add WattTime as a provider

### ISSUE THREAD

**Issue by mrchrisadams** [thegreenwebfoundation/grid-intensity-go#25]
We have the following providers:

- ✅ electricty map 
- ✅ember 
- ✅ carbon intensity (UK only) 

Wattime is a well known provider of marginal intensity information, and having them as a provider, would mean anyone using this library could expose metrics suitable for use their marginal carbon intensity figures for working out fairly high time resolution software carbon intensity scores.

This might be bit more complicated, as I can't remember if they support country codes - I think they use grid identifiers instead, which they refer to as balancing authorities.

You can fetch a list of the supported regions here, and we may need a way to do mapping between the two, or have a way to support listing geo coordinates, to let the mapping happen at watttime's end.

More docs below - we might want to have a separate call to think this one through together TBH, and spec it out properly before starting
https://www.watttime.org/api-documentation/#determine-grid-region

**Comment by rossf7** [comment]
See https://github.com/thegreenwebfoundation/grid-intensity-exporter/issues/1#issuecomment-1167193325 the higher time resolution of this data could be useful for autoscaling orchestrated systems like Nomad or Kubernetes.

Including Wattime support in https://github.com/thegreenwebfoundation/grid-intensity-exporter would enable some carbon aware scheduling. 

We can then look at more advanced integrations like adding to the experimental carbon branch for the Nomad scheduler https://github.com/hashicorp/nomad/blob/h-carbon-meta/CARBON.md

**Comment by rossf7** [comment]
Yes, for the first version I think we should use the simple "cheating" approach. I agree once we know more how users want to consume the WattTime data we can look at mapping co-ordinates to regions. 

Until then we can link to the WattTime docs in our docs and even in the help text for the CLI.
https://www.watttime.org/api-documentation/#determine-grid-region

## CLI arguments

From a CLI POV I think we should try to standardise on --provider and --region flags. This is similar to what we do in the exporter with the GRID_INTENSITY_PROVIDER and GRID_INTENSITY_REGION env vars.

So for WattTime an example command would be

```
grid-intensity --provider=watttime.org --region=CAISO_NORTH
```

For Ember we used `--country-code` but I'm thinking we should rename this to `--region`. 

In the future we might have providers that need both country and region params. Or country and grid params like in your example. But until then I think a single param is more intuitive. WDYT?

## Data

This is the raw data returned by the WattTime API for real time emissions.
https://www.watttime.org/api-documentation/#real-time-emissions-index

```
{
  "ba": "CAISO_NORTH",
  "freq": "300",
  "percent": "71",
  "moer": "888",
  "point_time": "2022-07-04T13:35:00Z"
}
```

> freq | Duration in seconds for which the data is valid from point_time
> percent | A percentile value between 0 (minimum MOER in the last month i.e. clean) and 100 (maximum MOER in the last month i.e. dirty) representing the relative realtime marginal emissions intensity.
> moer | Marginal Operating Emissions Rate (MOER) value measured in lbs/MWh.

A concern I have is the units for `moer`. I'd prefer to use gCO2/kWh like the other providers. I think a lot of users would expect the data to be in metric units :D so this could cause some confusion.

Should we convert the data? Or is that adding complexity? 

Also do we want to do any post processing on the data? e.g. rename `ba` to `balancing_authority`
This might make the data more readable but I'm not sure if its worth the effort as the data will differ between providers.

## Authentication

Similar to region I think we should ask users to register via the WattTime API.
https://www.watttime.org/api-documentation/#authentication

They can then provide their creds via env vars.

```
WATT_TIME_USERNAME=rossf7 \
WATT_TIME_PASSWORD=***** \
grid-intensity --provider=watttime.org --region=CAISO_NORTH
```

We can have a setting for the JWT token on the provider. If this is empty then we call the login endpoint.

https://www.watttime.org/api-documentation/#login-amp-obtain-token

The token expires after 30 mins. Once it expires fetching the data will return a 401. We can trap this error and try to get a new token by calling the login endpoint again.

This will be useful for the exporter and other long running API clients to reduce the number of API calls they need to make.

@mrchrisadams Thanks for your notes. I've been kicking the tires on the API and added my 2 centimos here.

I'm fine to continue discussing here but I can also write this up as a spec if that works better.

**Comment by rossf7** [comment]
Thanks Chris, using average instead of actual improves the naming. I'll also add the units. They are known as labels in prometheus lingo.

I'll implement these in the next PR that moves the exporter code.

**Comment by rossf7** [comment]
WattTime provider shipped in the 0.3.0 release
